src/proxy/dns.rs

```diff
@@ -0,0 +1,532 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::config;
+use crate::config::ProxyMode;
+use crate::proxy::{Error, ProxyInputs};
+use crate::service::Service;
+use crate::workload::{NetworkAddress, Workload};
+use istio_dns::forwarder;
+use istio_dns::name_util::trim_domain;
+use istio_dns::proxy::Proxy;
+use istio_dns::resolver::{Answer, Resolver};
+use itertools::Itertools;
+use log::warn;
+use std::collections::HashSet;
+use std::net::{IpAddr, SocketAddr};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+use tokio::net::{TcpListener, UdpSocket};
+use trust_dns_client::rr::LowerName;
+use trust_dns_proto::error::ProtoErrorKind;
+use trust_dns_proto::op::ResponseCode;
+use trust_dns_proto::rr::{Name, RData, Record, RecordType};
+use trust_dns_resolver::system_conf::read_system_conf;
+use trust_dns_server::authority::LookupError;
+use trust_dns_server::server::Request;
+use trust_dns_server::ServerFuture;
+
+const DEFAULT_TCP_REQUEST_TIMEOUT: u64 = 5;
+const DEFAULT_TTL_SECONDS: u32 = 30;
+
+/// A proxy that serves known hostnames from ztunnel data structures. Unknown hosts are
+/// forwarded to an upstream resolver.
+pub(super) struct DnsProxy {
+    addr: SocketAddr,
+    server: ServerFuture<Proxy>,
+}
+
+impl DnsProxy {
+    /// Creates a new proxy.
+    pub(super) async fn new(cfg: config::Config, pi: ProxyInputs) -> Result<Option<Self>, Error> {
+        // Create the DNS proxy only if configured to do so.
+        match cfg.enable_dns_proxy {
+            None | Some(false) => return Ok(None),
+            _ => Ok(Some(
+                Self::new_internal(cfg.dns_proxy_addr, cfg.proxy_mode, pi).await?,
+            )),
+        }
+    }
```

Comment:
Nit: I'd kinda prefer the cfg check to be moved out and live in `proxy.rs` rather than here (and then drop the `new`/`new_internal` stuff)

Reply:
Yeah I was going back and forth on this as well. I think my thought here was that I tried to keep as much DNS code as possible out of the other files. But it's a simple check ... I'm fine moving it.